test/basic_radix2_domain.cpp

```diff
@@ -0,0 +1,111 @@
+    const std::size_t fft_count = 5;
+    const std::array<std::size_t, fft_count> fft_sizes = {1 << 16, 1 << 17, 1 << 18, 1 << 19, 1 << 20};
+    std::array<std::vector<value_type>, fft_count> test_data;
+    std::chrono::time_point<std::chrono::high_resolution_clock> gen_start(std::chrono::high_resolution_clock::now());
+    for (std::size_t i = 0; i < fft_count; ++i) {
+        test_data[i].resize(fft_sizes[i]);
+        for (std::size_t j = 0; j < fft_sizes[i]; ++j) {
+            test_data[i][j] = nil::crypto3::algebra::random_element<FieldType>();
+        }
+    }
+    std::cout << "Generation: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                                       std::chrono::high_resolution_clock::now() - gen_start)
+                                       .count()
+              << " ms" << std::endl;
+
+    // manually calculate the power, saving all the intermediate powers
+    std::chrono::time_point<std::chrono::high_resolution_clock> cache_start(std::chrono::high_resolution_clock::now());
+    std::vector<std::shared_ptr<std::vector<value_type>>> omega_powers(fft_count);
+    for (std::size_t i = 0; i < fft_count; i++) {
+        const std::size_t logn = log2(fft_sizes[i]);
+        omega_powers[i].reset(new std::vector<value_type>);
+        omega_powers[i]->resize(fft_sizes[i]);
+        (*omega_powers[i])[0] = unity_root<FieldType>(fft_sizes[i]);
+        for (std::size_t j = 1; j < fft_sizes[i]; j++) {
+            (*omega_powers[i])[j] = (*omega_powers[i])[j - 1] * (*omega_powers[i])[0];
+        }
+    }
+    std::cout << "Cache: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                                  std::chrono::high_resolution_clock::now() - cache_start)
+                                  .count()
+              << " ms" << std::endl;
+
+    std::chrono::time_point<std::chrono::high_resolution_clock> start_fft(std::chrono::high_resolution_clock::now());
+    for (std::size_t i = 0; i < fft_count; ++i) {
+        nil::crypto3::math::detail::basic_radix2_fft<FieldType>(
+            test_data[i],
+            unity_root<FieldType>(fft_sizes[i])); //omega_powers[i]);
+    }
+
+    std::cout << "Uncached FFT: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                              std::chrono::high_resolution_clock::now() - start_fft)
+                              .count()
+              << " ms" << std::endl;
+
+    std::chrono::time_point<std::chrono::high_resolution_clock> start_cached(std::chrono::high_resolution_clock::now());
+    for (std::size_t i = 0; i < fft_count; ++i) {
+        nil::crypto3::math::detail::basic_radix2_fft<FieldType>(
+            test_data[i],
+            unity_root<FieldType>(fft_sizes[i]),
+            omega_powers[i]);
+    }
+    std::cout << "Cached FFT: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                              std::chrono::high_resolution_clock::now() - start_cached)
```

Comment:
Can you move this a few tabs to the left?

Reply:
Re-tabbed all of them.